Timeline containing exchange or named error events

alexjeffburke opened the pull request:
@papandreou I managed to get a branch working that has some kind of split between the exchanges it sees and the asserting portion - it uses named errors for the various cases and asserts/throws as appropriate. It passes the entire test suite so pushing as I'm really keen to see what you think.

papandreou commented on lib/unexpectedMitm.js:
```diff
@@ -1151,28 +1157,55 @@ module.exports = {
+                        reject(e);
+                    });
+                }).catch(function () {
+                    var lastPartOrError = exchangeParts[exchangeParts.length - 1];
```
Will `exchangeParts.length` always be > 0 here?

alexjeffburke replied:
Good question - I'll instrument the code to check. This might be relying on the return undefined behaviour but if so I think an explicit check might be better.

papandreou replied:
I realize that it'll still come out right even when dereferencing `-1` because of the truthiness check below, but I'd rather not go out of bounds. Call me a freak :)

alexjeffburke replied:
The diff changed but it's addressed this yet - this is taken care of in the work that pushes this in the direction of separated mocker.

papandreou commented on lib/unexpectedMitm.js:
```diff
@@ -1047,22 +1052,30 @@ module.exports = {
 
                         function assertMockResponse(mockResponse, mockResponseError) {
                             var mockRequest = new messy.HttpRequest(requestProperties);
+                            var httpPart = {
```
Not sure I understand the `httpPart` variable name, what does it mean? :)

alexjeffburke replied:
Yeah, I was using `part` to mean a portion of the http conversation - at minimum to pe consistent with the rest of the file it wants the http prefix taken off. But I'm open for naming suggestions.. :D

papandreou commented on lib/unexpectedMitm.js:
```diff
@@ -1047,22 +1052,30 @@ module.exports = {
+                                spec: undefined
+                            };
+
                             var httpExchange = new messy.HttpExchange({request: mockRequest});
```
I think I'd set this directly as the `exchange` property above, but it's certainly a matter of taste.

papandreou commented on lib/unexpectedMitm.js:
```diff
@@ -876,24 +893,18 @@ module.exports = {
                 }
                 var nextRequestDescriptionIndex = 0;
 
+                var exchangeParts = [];
```
Can we find a better variable name that makes it obvious that it'll contain a mix of `{ spec, httpExchange }` objects and errors? Maybe something with `timeline`?

alexjeffburke replied:
Right, yes - at minimum it's not actually exchangeParts, it's more like conversationParts. But that works for me too - think part of the difficulty is this is a kind of checkpoint rather than where we might want it to ultimately end up.

alexjeffburke commented:
@papandreou I’ve rebased this and cleaned it up a tad, getting more and more convince this is the right direction :)

What I’ve also done is start an experiment atop eats here that makes the mocking code something that can be instantiated separately - I’m very close to having early exit pop out a diff after the fact. Will see how far I can get.

papandreou commented on lib/unexpectedMitm.js:
```diff
@@ -875,327 +597,26 @@ module.exports = {
-                        });
-                    }));
+                var mocker = new UnexpectedMitmMocker({
+                    descriptions: requestDescriptions
```
Could this either be `descriptions` or `requestDescriptions` in both places? I think I'd prefer `requestDescriptions`. Or how about `expectedExchanges`?

alexjeffburke replied:
Agreed the inconsistency doesn't make much sense.

As for the specific name, I think part of the difficult is it's the expected request but an arranged response. I think I'll go with `requestDescriptions` for now and if we iterate we can do another pass and change it throughout.

papandreou commented on lib/UnexpectedMitmMocker.js:
```diff
@@ -0,0 +1,538 @@
+                sawHeaders = true;
+            }
+
+            match = chunkString.match(/^Content-Length: (\d+)/m);
```
Regexp should be case insensitive (`/mi`)

papandreou commented on lib/UnexpectedMitmMocker.js:
```diff
@@ -0,0 +1,538 @@
+            var chunkString = chunk.toString();
+            var match;
+
+            match = chunkString.match(/^Transfer-Encoding: /m);
```
Regexp should be case insensitive (`/mi`)

papandreou commented on lib/UnexpectedMitmMocker.js:
```diff
@@ -0,0 +1,538 @@
+    this.fulfilmentValue = null;
+
+    var that = this;
+    Object.defineProperty(this, 'isComplete', {
```
Couldn't this live in `UnexpectedMitmMocker.prototype`?

alexjeffburke replied:
IIRC it's actually not used right now - my temptation would be to remove it until necessary (for some additional context, this was an early version of what lead to the completion marker).

papandreou replied:
Even better :)

papandreou commented on lib/UnexpectedMitmMocker.js:
```diff
@@ -0,0 +1,538 @@
+
+                timeline.push({
+                    exchange: httpExchange,
+                    spec: eventSpec
```
Could we find a way to call the property and the variables the same? Would be nice for consistency, and we'll be ready for using the shorthand syntax later :)

papandreou commented on lib/UnexpectedMitmMocker.js:
```diff
@@ -0,0 +1,538 @@
+        try {
+            var consumerResult = consumptionFunction();
+            // ensure consumption function returned a promise
+            consumer = Promise.resolve().then(function () {
```
Wouldn't `Promise.resolve(consumerResult)` work?
